File: Medium/2026-09-10-2265-Count-Nodes-Equal-to-Average-of-Subtree/solution.java
class Solution {
    private int count = 0;

    public int averageOfSubtree(TreeNode root) {
        calculateSumAndCount(root);
        return count;
    }

    private int[] calculateSumAndCount(TreeNode node) {
        if (node == null) {
            return new int[]{0, 0}; // sum, count
        }
        
        int[] left = calculateSumAndCount(node.left);
        int[] right = calculateSumAndCount(node.right);
        
        int sum = left[0] + right[0] + node.val;
        int count = left[1] + right[1] + 1;
        
        if (node.val == sum / count) {
            count++;
        }
        
        return new int[]{sum, count};
    }
}

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}

File: Medium/2026-09-10-2265-Count-Nodes-Equal-to-Average-of-Subtree/solution.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val === undefined ? 0 : val)
 *     this.left = (left === undefined ? null : left)
 *     this.right = (right === undefined ? null : right)
 * }
 */

/**
 * @param {TreeNode} root
 * @return {number}
 */
var averageOfSubtree = function(root) {
    let count = 0;

    const dfs = (node) => {
        if (!node) return [0, 0]; // [sum, count]

        const [leftSum, leftCount] = dfs(node.left);
        const [rightSum, rightCount] = dfs(node.right);

        const totalSum = leftSum + rightSum + node.val;
        const totalCount = leftCount + rightCount + 1;
        const average = Math.floor(totalSum / totalCount);

        if (average === node.val) {
            count++;
        }

        return [totalSum, totalCount];
    };

    dfs(root);
    return count;
};